### PR TITLE
CAPTCHA checking before posting

### PR DESCRIPTION
**Pull request by MacilPrime**
This change makes it so when you cache a CAPTCHA with shift+enter, the CAPTCHA is validated before being cached. This solves the issue where you make a post, wait a while for it and maybe an image to upload, just to get told the CAPTCHA was bad and you have to do the upload again.

This behavior is optional and controlled by the new "Validate CAPTCHA" option, which defaults to on. If there are any connection errors in checking the CAPTCHA, then the CAPTCHA is cached anyway.

Uncached CAPTCHAs aren't checked, partly because it would be more difficult (the ajax check would have to happen in the submit handler, and queue the rest of the submission to happen later) and because if the connection to google's recaptcha server is slow for the user, the validation will slow down posting for the user.

It works by using the recaptcha's noscript fallback. If you don't have javascript turned on, then 4chan shows an iframe to a recaptcha page with a captcha, and when you solve it, it gives you a new challenge code to copy and paste into a field on the 4chan posting form. This allows the captcha validation to be decoupled from the actual posting.

Google's recaptcha server is a different domain from 4chan, so the ajax call to it is cross-domain, but it doesn't set its headers to explicitly allow that. Chrome allows userscripts to do cross-domain ajax calls by default. For Firefox Greasemonkey/Scriptish users, GM_xmlhttpRequest is used. I made a function $.crossAjax which acts as a wrapper for both of these. I haven't tested it in Opera.

### CODE REVIEW

File: script.coffee [hunk 1]
```diff
@@ -336,6 +337,35 @@ $.extend $,
     r.withCredentials = true if type is 'post'
     r.send form
     r
+  crossAjax: (url, callbacks, opts={}) ->
+    if typeof GM_xmlhttpRequest != "undefined"
+      console.log "Using GM_xmlhttpRequest"
```
Comment (ihavenoface) [inline script.coffee]: Throw out those logs.

You can also use `$.log` in case you need logging.

File: script.coffee [hunk 2]
```diff
@@ -336,6 +337,35 @@ $.extend $,
...
+        gmopts.onabort = callbacks.onabort
+      if callbacks.onerror
+        gmopts.onerror = callbacks.onerror
+      GM_xmlhttpRequest gmopts
+    else
+      console.log "Using regular $.ajax"
+      if callbacks.onload
+        newonload = ->
+          callbacks.onload
+            readyState: @readyState
```
Comment (ihavenoface) [inline script.coffee]:
```
readyState:   @readyState
responseText: @responseText
status:       @status
statusText:   @statusText
```

Same for `gmopts =`, etc..

File: script.coffee [hunk 3]
```diff
@@ -2019,18 +2049,50 @@ QR =
...
+      callbacks =
+        onload: (data) ->
+          doc = d.implementation.createHTMLDocument ''
+          doc.documentElement.innerHTML = data.responseText
+          if ta = $ 'textarea', doc
+            key = ta.innerHTML
+            console.log "Good CAPTCHA"
+            QR.cleanError()
+            QR.captcha.addCaptcha key, "manual_challenge"
+          else if $ '#recaptcha_response_field', doc
```
Comment (ihavenoface) [inline script.coffee:2069]: Can use `$.id 'recaptcha_response_field', doc` instead.

**Comment by ihavenoface** [comment]
Very nice idea and the implementation looks solid so far. Thank you.

File: script.coffee [hunk 4]
```diff
@@ -336,6 +337,35 @@ $.extend $,
     r.withCredentials = true if type is 'post'
     r.send form
     r
+  crossAjax: (url, callbacks, opts={}) ->
+    if typeof GM_xmlhttpRequest != "undefined"
```
Comment (ihavenoface) [inline script.coffee]: I would prefer not having double quotes where not necessary, i.e. `'undefined'`.

File: script.coffee [hunk 5]
```diff
@@ -2019,18 +2046,47 @@ QR =
...
+          recaptcha_response_field: response
+      callbacks =
+        onload: (data) ->
+          doc = d.implementation.createHTMLDocument ''
+          doc.documentElement.innerHTML = data.responseText
+          if ta = $ 'textarea', doc
+            key = ta.innerHTML
+            QR.cleanError()
+            QR.captcha.addCaptcha key, 'manual_challenge'
+          else if $.id 'recaptcha_response_field', doc
```
Comment (ihavenoface) [inline script.coffee]: You can revert this. `$.id` only accepts one argument as of now. My bad.

File: script.coffee [hunk 6]
```diff
@@ -336,6 +337,28 @@ $.extend $,
     r.withCredentials = true if type is 'post'
     r.send form
     r
+  crossAjax: (url, callbacks, opts={}) ->
+    if typeof GM_xmlhttpRequest != 'undefined'
```
Comment (vendethiel) [inline script.coffee:341]: `if GM_xmlhttpRequest?`

Comment (vendethiel) [inline script.coffee:341]: @AgentME yep, that's the soak

Comment (ihavenoface) [inline script.coffee:341]: Why are you pinging random people?

Comment (vendethiel) [inline script.coffee:341]: Am not.
proof

Comment (ihavenoface) [inline script.coffee:341]: Oh alright.

**Pull request merged**